go-controller/pkg/ovn/egressip.go

```diff
@@ -123,14 +128,24 @@ func (oc *Controller) reconcileEgressIP(old, new *egressipv1.EgressIP) (err erro
 		}
 	}
 
-	// Check the old egress IP and its status assignments. This is mainly done
-	// if a node is detected as not ready/reachable: we'll end up getting called
-	// from deleteEgressNode, which will have updated the object with only the
-	// statuses to keep. Hence, if the new object does not have the old
-	// assignment, we need to remove those.
+	// Check the old egress IP and its status assignments. When an EgressIP is
+	// deleted we'll need to clear all old status assignments. If a user removes
+	// an egress IP from .spec.egressIPs we however need to guard against adding
+	// the same status to statusToRemove twice, which would happen with the code
+	// above for validStatuses because both newEIP and oldEIP will have the old
+	// assignment status.
 	for _, oldStatus := range oldEIP.Status.Items {
-		if !validStatusIPs.Has(oldStatus.EgressIP) {
-			statusToRemove = append(statusToRemove, oldStatus)
+		if !validSpecIPs.Has(oldStatus.EgressIP) {
+			exists := false
+			for _, status := range statusToRemove {
+				if reflect.DeepEqual(status, oldStatus) {
+					exists = true
```

Comment:
NIT: can we just move `statusToRemove = append(statusToRemove, oldStatus)` here and not use `exists`?

Reply:
Not really because of what I mentioned in the doc bloc above:
```
        // ....If a user removes
	// an egress IP from .spec.egressIPs we however need to guard against adding
	// the same status to statusToRemove twice, which would happen with the code
	// above for validStatuses because both newEIP and oldEIP will have the old
	// assignment status.
```

If an update comes in: 

old: 

```yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 10.0.208.210
    - 10.0.208.211
    namespaceSelector:
        matchLabels:
            name: test
status:
    items:
    - EgressIP: 10.0.208.210
      Node: node1
    - EgressIP: 10.0.208.211
      Node: node2
```

new: 

```yaml
apiVersion: k8s.ovn.org/v1
kind: EgressIP
metadata:
    name: egressip
spec:
    egressIPs:
    - 10.0.208.210
    namespaceSelector:
        matchLabels:
            name: test
status:
    items:
    - EgressIP: 10.0.208.210
      Node: node1
    - EgressIP: 10.0.208.211
      Node: node2
```

In which case, without this check: `statusToRemove` would contain duplicate items for `    - EgressIP: 10.0.208.211
      Node: node2`

Reply:
sets can only be comprised of comparable primitive types, and I was going to say that this applies to map keys as well..but I learned something new today: a map can be comprised of a struct if all the struct fields are comparable, this is the case for an `EgressIPStatusItem`, so I will change this  to a  map based implementation.